istioctl/pkg/auth/validator.go

```diff
@@ -0,0 +1,122 @@
+	serviceRoleBindings  []model.Config
+	Report               strings.Builder
+}
+
+const (
+	RoleNotFound  = "serviceRoleNotFound: %q used by ServiceRoleBinding %q at namespace %q.\n"
+	RoleNotUsed   = "serviceRoleNotUsed: ServiceRole %q at namespace %q.\n"
+	PolicyValid   = "Authorization policy is valid.\n"
+	PolicyMissing = "needs to have both ServiceRole and ServiceRoleBinding for validation"
+)
+
+// CheckAndReport checks for Istio authentication and authorization mis-usage.
+func (v *Validator) CheckAndReport() error {
+	err := v.getRoleAndBindingLists()
+	if err != nil {
+		return err
+	}
+	v.CheckAndReportRBAC()
+	if v.Report.String() == "" {
+		v.Report.WriteString(GetPolicyValidReport())
+	}
+	return nil
+}
+
+func (v *Validator) CheckAndReportRBAC() {
+	usedRoleNames := map[string]bool{}
+	// Check if ServiceRoleBinding is using an non existent ServiceRole.
+	for _, binding := range v.serviceRoleBindings {
+		bindingSpec := binding.Spec.(*rbacproto.ServiceRoleBinding)
+		namespace := binding.Namespace
+		roleName := bindingSpec.RoleRef.Name
+		if v.doesRoleExist(namespace, roleName) {
+			roleKey := getRoleKey(namespace, roleName)
+			usedRoleNames[roleKey] = true
+		} else {
+			v.Report.WriteString(GetRoleNotFoundReport(roleName, binding.Name, namespace))
+		}
+	}
+	// Check if ServiceRole is unused.
+	for roleKey := range v.RoleKeyToServiceRole {
+		if _, found := usedRoleNames[roleKey]; !found {
+			namespace, roleName := getNamespaceAndRoleNameFromRoleKey(roleKey)
+			v.Report.WriteString(GetRoleNotUsedReport(roleName, namespace))
+		}
+	}
+}
+
+// doesRoleExist check if a role exist in the given namespace in the provided policy file.
+func (v *Validator) doesRoleExist(namespace, roleName string) bool {
+	roleKey := getRoleKey(namespace, roleName)
+	if _, found := v.RoleKeyToServiceRole[roleKey]; found {
+		return true
+	}
+	return false
+}
+
+// getRoleKey joins namespace and role name with a forward slash and returns the result.
+func getRoleKey(namespace, roleName string) string {
+	return fmt.Sprintf("%s/%s", namespace, roleName)
+}
+
+// getNamespaceAndRoleNameFromRoleKey returns namespace and role name from the given role key.
+func getNamespaceAndRoleNameFromRoleKey(roleKey string) (string, string) {
+	foo := strings.Split(roleKey, "/")
+	return foo[0], foo[1]
+}
+
+// getRoleAndBindingLists get roles and bindings from the provided files to the appropriate data structures for the validator.
+func (v *Validator) getRoleAndBindingLists() error {
+	configsFromFiles, err := getConfigsFromFiles(v.PolicyFiles)
+	if err != nil {
+		return err
+	}
+	for _, role := range configsFromFiles[model.ServiceRole.Type] {
+		roleKey := getRoleKey(role.Namespace, role.Name)
+		v.RoleKeyToServiceRole[roleKey] = role
+	}
+	v.serviceRoleBindings = configsFromFiles[model.ServiceRoleBinding.Type]
+	if len(v.serviceRoleBindings) == 0 || len(v.RoleKeyToServiceRole) == 0 {
+		return fmt.Errorf(PolicyMissing)
```

Comment:
It is ok if there is no role or binding defined. Why is it an error?

Reply:
Since this validator doesn't have a view of the world, i.e. it only knows what presents in the input files. So if you have a ServiceRole with no ServiceRoleBinding, that ServiceRole is considered not being used, since there is no ServiceRoleBinding.

Reply:
There are two cases:
1. If no RBAC policy exists, it should be a valid state. 
2. If there is ServiceRole but no ServiceRoleBinding, or vise versa, we should report error. But the error message should specifically mention which policy is missing (role or binding).

Reply:
Thanks. I will make the change

Reply:
Done.